Avoid calling FrameSelection::updateAppearanceAfterLayout() when performing post-layout tasks

rniwa opened the pull request:
#### 44856812df2915e4c0236ed677fdeb485b50bf5b
<pre>
Avoid calling FrameSelection::updateAppearanceAfterLayout() when performing post-layout tasks
<a href="https://bugs.webkit.org/show_bug.cgi?id=247041">https://bugs.webkit.org/show_bug.cgi?id=247041</a>

Reviewed by Wenson Hsieh.

Removed the calls to scheduleAppearanceUpdateAfterStyleChange in Document::resolveStyle
and updateAppearanceAfterLayout in LocalFrameView::performPostLayoutTasks.

Now every appearance update happens with the rendering update in Page::doAfterUpdateRendering.

Fixed &amp; rebaselined the various tests to account for this delayed update of selection &amp; focus.

* LayoutTests/accessibility/mac/focus-setting-selection-syncronizing-not-clearing-expected.txt:
* LayoutTests/accessibility/mac/selection-boundary-userinfo.html:
* LayoutTests/accessibility/mac/selection-change-userinfo.html:
* LayoutTests/accessibility/mac/selection-element-tabbing-to-link.html:
* LayoutTests/accessibility/mac/selection-value-changes-for-aria-textbox-expected.txt:
* LayoutTests/editing/input/caret-at-the-edge-of-input.html:
* LayoutTests/fast/forms/autofocus-opera-003.html:
* LayoutTests/fast/forms/input-text-scroll-left-on-blur.html:
* LayoutTests/fast/forms/password-scrolled-after-caps-lock-toggled.html:
* LayoutTests/fast/forms/scroll-into-view-and-show-validation-message.html:
* LayoutTests/fast/repaint/4776765.html:
* LayoutTests/fast/repaint/selection-gap-absolute-child.html:
* LayoutTests/fast/repaint/selection-gap-flipped-absolute-child.html:
* LayoutTests/fast/repaint/selection-gap-transformed-absolute-child.html:
* LayoutTests/fast/repaint/selection-gap-transformed-fixed-child.html:
* LayoutTests/fast/repaint/selection-ruby-rl.html:
* LayoutTests/fast/repaint/text-selection-overflow-hidden.html:
* LayoutTests/fast/text/incorrect-deselection-across-multiple-elements.html:
* LayoutTests/platform/mac-wk1/fast/repaint/4776765-expected.txt:

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::revealSelectionIfNeededAfterLoadingImageForElement): Invalidate the caret rect
since the act of loading an image may have shifted the caret. Without this change,
TestWebKitAPI.PasteImage.RevealSelectionAfterPastingImage will fail.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::FrameSelection):
(WebCore::FrameSelection::setSelection):
(WebCore::FrameSelection::willBeRemovedFromFrame):
(WebCore::FrameSelection::updateAppearanceAfterUpdatingRendering): Renamed from
updateAppearanceAfterLayoutOrStyleChange.
(WebCore::FrameSelection::updateAppearanceAfterLayout): Deleted.
(WebCore::FrameSelection::scheduleAppearanceUpdateAfterStyleChange): Deleted.
(WebCore::FrameSelection::appearanceUpdateTimerFired): Deleted.

* Source/WebCore/editing/FrameSelection.h:

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::performPostLayoutTasks):

* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):

Canonical link: <a href="https://commits.webkit.org/262691@main">https://commits.webkit.org/262691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee663495839b962599d4a888dda9535eea0c0f6f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2287 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/2307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2396 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3220 "Failed to checkout and rebase branch from PR 12397") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2319 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2262 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/2396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2370 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/3220 "Failed to checkout and rebase branch from PR 12397") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2308 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/2396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2059 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3074 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/2396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2041 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2118 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2089 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/3232 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2100 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2029 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2047 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2223 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/258 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3796,12 +3796,6 @@ void LocalFrameView::performPostLayoutTasks()
     LOG(Layout, "LocalFrameView %p performPostLayoutTasks", this);
     updateHasReachedSignificantRenderedTextThreshold();
 
-    if (auto& selection = m_frame->selection(); selection.isFocusedAndActive()) {
-        // FIXME (247041): We should be able to remove this appearance update altogether,
-        // and instead defer updates until the next rendering update.
-        selection.updateAppearanceAfterLayout();
-    }
-
```
🎉

whsieh commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2570,24 +2566,7 @@ void FrameSelection::setShouldShowBlockCursor(bool shouldShowBlockCursor)
-}
-
-void FrameSelection::updateAppearanceAfterLayoutOrStyleChange()
+void FrameSelection::updateAppearanceAfterUpdateRendering()
```
Nit - maybe, `updateAppearanceAfterUpdatingRendering`, or `updateAppearanceAfterRenderingUpdate`?

rniwa replied:
Okay, I'd go with updateAppearanceAfterUpdatingRendering.

whsieh commented on LayoutTests/accessibility/mac/selection-value-changes-for-aria-textbox-expected.txt:
```diff
@@ -10,7 +10,5 @@ PASS successfullyParsed is true
 TEST COMPLETE
 Successfully received AXSelectedTextChanged
 Successfully received AXSelectedTextChanged
-Successfully received AXSelectedTextChanged
-Successfully received AXSelectedTextChanged
```
I recall that when I last wrote a (very similar) version of this patch a few months ago, I was unsure about just rebaseling this test, since it seemed like a potentially risky behavior change. But thinking about this again, maybe it's okay since we're just batching together these notifications, and a real AX user doesn't need to know about each individual selection change/editing operation in a series of selection changes/editing that happen in the same frame.

rniwa replied:
Right. That was my reasoning as well.

rniwa committed:
Avoid calling FrameSelection::updateAppearanceAfterLayout() when performing post-layout tasks
https://bugs.webkit.org/show_bug.cgi?id=247041

Reviewed by Wenson Hsieh.

Removed the calls to scheduleAppearanceUpdateAfterStyleChange in Document::resolveStyle
and updateAppearanceAfterLayout in LocalFrameView::performPostLayoutTasks.

Now every appearance update happens with the rendering update in Page::doAfterUpdateRendering.

Fixed & rebaselined the various tests to account for this delayed update of selection & focus.

* LayoutTests/accessibility/mac/focus-setting-selection-syncronizing-not-clearing-expected.txt:
* LayoutTests/accessibility/mac/selection-boundary-userinfo.html:
* LayoutTests/accessibility/mac/selection-change-userinfo.html:
* LayoutTests/accessibility/mac/selection-element-tabbing-to-link.html:
* LayoutTests/accessibility/mac/selection-value-changes-for-aria-textbox-expected.txt:
* LayoutTests/editing/input/caret-at-the-edge-of-input.html:
* LayoutTests/fast/forms/autofocus-opera-003.html:
* LayoutTests/fast/forms/input-text-scroll-left-on-blur.html:
* LayoutTests/fast/forms/password-scrolled-after-caps-lock-toggled.html:
* LayoutTests/fast/forms/scroll-into-view-and-show-validation-message.html:
* LayoutTests/fast/repaint/4776765.html:
* LayoutTests/fast/repaint/selection-gap-absolute-child.html:
* LayoutTests/fast/repaint/selection-gap-flipped-absolute-child.html:
* LayoutTests/fast/repaint/selection-gap-transformed-absolute-child.html:
* LayoutTests/fast/repaint/selection-gap-transformed-fixed-child.html:
* LayoutTests/fast/repaint/selection-ruby-rl.html:
* LayoutTests/fast/repaint/text-selection-overflow-hidden.html:
* LayoutTests/fast/text/incorrect-deselection-across-multiple-elements.html:
* LayoutTests/platform/mac-wk1/fast/repaint/4776765-expected.txt:

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::revealSelectionIfNeededAfterLoadingImageForElement): Invalidate the caret rect
since the act of loading an image may have shifted the caret. Without this change,
TestWebKitAPI.PasteImage.RevealSelectionAfterPastingImage will fail.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::FrameSelection):
(WebCore::FrameSelection::setSelection):
(WebCore::FrameSelection::willBeRemovedFromFrame):
(WebCore::FrameSelection::updateAppearanceAfterUpdatingRendering): Renamed from
updateAppearanceAfterLayoutOrStyleChange.
(WebCore::FrameSelection::updateAppearanceAfterLayout): Deleted.
(WebCore::FrameSelection::scheduleAppearanceUpdateAfterStyleChange): Deleted.
(WebCore::FrameSelection::appearanceUpdateTimerFired): Deleted.

* Source/WebCore/editing/FrameSelection.h:

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::performPostLayoutTasks):

* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):

Canonical link: https://commits.webkit.org/262691@main
